cpp/include/cudf/detail/iterator.cuh

```diff
@@ -0,0 +1,207 @@
+  Element const null_replacement{};     ///< value returned when element is null
+
+/** -------------------------------------------------------------------------*
+ * @brief constructor
+ * @param[in] _col column device view of cudf column
+ * @param[in] null_replacement The value to return for null elements
+ * -------------------------------------------------------------------------**/
+  null_replaced_value_accessor(column_device_view const& _col, Element null_val)
+    : col{_col}, null_replacement{null_val}
+  {
+    // verify valid is non-null, otherwise, is_valid() will crash
+    CUDF_EXPECTS(_col.nullable(), "Unexpected non-nullable column.");
+  }
+
+  CUDA_DEVICE_CALLABLE
+  Element operator()(cudf::size_type i) const {
+    return col.is_valid_nocheck(i) ?
+      col.element<Element>(i) : null_replacement;
+  }
+};
+
+/** -------------------------------------------------------------------------*
+ * @brief value accessor of column without null bitmask
+ * A unary functor returns scalar value at `id`.
+ * `operator() (cudf::size_type id)` computes `element`
+ * This functor is only allowed for non-nullable columns.
+ *
+ * the return value for element `i` will return `column[i]`
+ *
+ * @throws `cudf::logic_error` if the column is nullable.
+ *
+ * @tparam Element The type of elements in the column
+ * -------------------------------------------------------------------------**/
+
+template <typename Element>
+struct column_device_view::value_accessor
+{
+  column_device_view const col;         ///< column view of column in device
+
+/** -------------------------------------------------------------------------*
+ * @brief constructor
+ * @param[in] _col column device view of cudf column
+ * -------------------------------------------------------------------------**/
+  value_accessor(column_device_view const& _col)
+    : col{_col}
+  {
+    // verify column does not have nulls
+    CUDF_EXPECTS(!_col.has_nulls(), "Unexpected column with nulls.");
+  }
+
+  CUDA_DEVICE_CALLABLE
+  Element operator()(cudf::size_type i) const {
+    return col.element<Element>(i);
+  }
+};
+
+template <typename T>
+column_device_view::const_iterator<T> column_device_view::begin() const {
+  CUDF_EXPECTS(data_type(experimental::type_to_id<T>()) == type(), "the data type mismatch");
+  return column_device_view::const_iterator<T>{
+      column_device_view::count_it{0},
+      column_device_view::value_accessor<T>{*this}};
+}
+template <typename T>
+column_device_view::const_iterator<T> column_device_view::end() const {
+  CUDF_EXPECTS(data_type(experimental::type_to_id<T>()) == type(), "the data type mismatch");
+  return column_device_view::const_iterator<T>{
+      column_device_view::count_it{size()},
+      column_device_view::value_accessor<T>{*this}};
+}
+
+template <typename T>
+column_device_view::const_null_iterator<T> column_device_view::nbegin(const T& null_val) const {
```

Comment:
I don't think these should be member functions since they are non-standard unlike `begin/end`. Furthermore, by having `nbegin/nend` both accept a `null_val` argument, it allows for the possibility of erroneously passing a different value into each.

This functionality should be kept as an explicit function call in `make_null_replacement_iterator`.

Reply:
removed nbegin, nend.